### PR TITLE
[Clang] strengthen checks for 'main' function to meet [basic.start.main] p3 requirements

### LINKED ISSUE #101512: [Clang] Diagnose `extern "C++" int main()`
This currently just seems to be ignored (the name is not mangled).  
It is explicitly disallowed by https://eel.is/c++draft/basic.start.main#3.sentence-5

> The `main` function shall not be declared with a *linkage-specification*.

This includes

```cpp
extern "C++" { int main(); }
```

This is also technically ill-formed

```cpp
extern "C" int main();
```

but it seems to be used in some places:

https://github.com/llvm/llvm-project/blob/98e4413a38f286147b863a6ead9625228ab0ec7d/libc/startup/linux/do_start.cpp#L23
https://github.com/llvm/llvm-project/blob/98e4413a38f286147b863a6ead9625228ab0ec7d/libc/startup/gpu/nvptx/start.cpp#L15

So this should probably just be a warning, not an error.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #101512

### CODE REVIEW

File: clang/test/SemaCXX/linkage1.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  extern "C" void main(); // expected-error {{invalid linkage specification 'extern "C"'}}
+}
+extern "C" {
+  int main(); // expected-error {{invalid linkage specification 'extern "C"'}}
+}
+extern "C" int main(); // expected-error {{invalid linkage specification 'extern "C"'}}
+extern "C" struct A { int main(); }; // ok
+
+namespace cpp {
+  extern "C++" int main(); // expected-error {{invalid linkage specification 'extern "C++"'}}
```
Comment (MitalAshok) [inline clang/test/SemaCXX/linkage1.cpp:13]: This one is allowed, https://eel.is/c++draft/basic.start.main#3.4 only disallows `extern "C"` `main`. If you are implementing that, could you add tests for functions (`extern "C" { int main(); }`) and variables (`extern "C" { void* main; }`) named `main`?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -990,6 +990,9 @@ def warn_main_redefined : Warning<"variable named 'main' with external linkage "
     "has undefined behavior">, InGroup<Main>;
 def ext_main_used : Extension<
     "referring to 'main' within an expression is a Clang extension">, InGroup<Main>;
+def err_invalid_linkage_specification : Extension<
+  "invalid linkage specification "
+  "'extern \"%select{C|C++}0\"'">;
```
Comment (MitalAshok) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:995]: This doesn't say why it's invalid. `'main' cannot have linkage specification 'extern "C(++)"'`?

Comment (a-tarasyuk) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: @MitalAshok I've updated the diagnostic message

File: clang/test/SemaCXX/linkage1.cpp (hunk 1 above)
Comment (a-tarasyuk) [reply clang/test/SemaCXX/linkage1.cpp]: @MitalAshok Thanks for the feedback. I've added [linkage3.cpp](https://github.com/llvm/llvm-project/pull/101853/files#diff-dc2bba9678021106302e8711c2140ad1e575de50b157537e64b37621e97b9ec9) to cover this case.

File: clang/lib/Sema/SemaDecl.cpp [hunk 3]
```diff
@@ -7353,6 +7353,15 @@ void emitReadOnlyPlacementAttrWarning(Sema &S, const VarDecl *VD) {
   }
 }
 
+static bool isMainVar(DeclarationName Name, VarDecl *VD) {
+  if (Name.getAsIdentifierInfo() && Name.getAsIdentifierInfo()->isStr("main") &&
+      !VD->getDescribedVarTemplate()) {
+    const DeclContext *DC = VD->getDeclContext();
+    return DC->getRedeclContext()->isTranslationUnit() || DC->isLinkageSpec();
```
Comment (MitalAshok) [inline clang/lib/Sema/SemaDecl.cpp:7360]: This doesn't seem to catch `extern "C" { namespace NS { int main; } }`

```suggestion
    return DC->getRedeclContext()->isTranslationUnit() ||
           VD->getLanguageLinkage() == CLanguageLinkage;
```

**Review by MitalAshok — CHANGES_REQUESTED** [review]
You can put some of the tests in `clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp`

You lost the check for `extern "C++" int main() {}` at global scope. To recap:

```c++
// These are disallowed by [basic.start.main]p3
//   The main function shall not be declared with a linkage-specification.
// We only want a pedantic warning for these
extern "C++" {
int main();
}
extern "C" {
int main();
}
// These are disallowed by [basic.start.main]p(3.4)
// We want a hard error
namespace X { extern "C" int main; }
extern "C" {
namespace Y {
  int main;
}
namespace Z {
  void main();
}
}
// These are allowed
namespace W {
  extern "C++" int main();
  extern "C" {
    extern "C++" {
      int main(void*);
    }
  }
}
```

(You should add a test for `extern "C" { namespace namespace_name { int main; } }` too)

Did you try doing it like https://github.com/llvm/llvm-project/issues/101512#issuecomment-2263650171 , checking `Decl->getLanguageLinkage()` rather than checking the DeclContext it's declared in?

File: clang/lib/AST/Decl.cpp [hunk 4]
```diff
@@ -3292,11 +3292,9 @@ bool FunctionDecl::isImmediateFunction() const {
...
 
 bool FunctionDecl::isMain() const {
-  const TranslationUnitDecl *tunit =
-    dyn_cast<TranslationUnitDecl>(getDeclContext()->getRedeclContext());
-  return tunit &&
-         !tunit->getASTContext().getLangOpts().Freestanding &&
-         isNamed(this, "main");
+  const DeclContext *DC = getDeclContext();
+  return isNamed(this, "main") && !getLangOpts().Freestanding &&
+         (DC->getRedeclContext()->isTranslationUnit() || DC->isLinkageSpec());
```
Comment (Sirraide) [inline clang/lib/AST/Decl.cpp:3297]:
```suggestion
         (DC->getRedeclContext()->isTranslationUnit() || isExternC());
```

I think this should do the trick. Also please add a test for
```c++
namespace foo { 
    extern "C++" void main() {} 
}
```
which is horrible, but valid.

File: clang/include/clang/Sema/Sema.h [hunk 5]
```diff
@@ -3585,8 +3585,9 @@ class Sema final : public SemaBase {
   /// \param OldT The portion of the type of the old declaration to check.
   bool canFullyTypeCheckRedeclaration(ValueDecl *NewD, ValueDecl *OldD,
                                       QualType NewT, QualType OldT);
-  void CheckMain(FunctionDecl *FD, const DeclSpec &D);
+  void CheckMain(FunctionDecl *FD, DeclContext *DC, const DeclSpec &D);
```
Comment (Sirraide) [inline clang/include/clang/Sema/Sema.h:3588]: You can get the decl context from the function decl, so I don’t think this change is necessary.

File: clang/test/SemaCXX/linkage1.cpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -pedantic-errors %s
+
```
Comment (Sirraide) [inline clang/test/SemaCXX/linkage1.cpp:2]: I think this was already pointed out, but some tests w/ `extern "C++"` would be nice, as well as tests for when a namespace is contained in a linkage-spec-decl.

File: clang/lib/Sema/SemaDecl.cpp (hunk 3 above)
Comment (Sirraide) [reply clang/lib/Sema/SemaDecl.cpp:7360]: Just `VD->isExternC()` instead of `VD->getLanguageLinkage() == CLanguageLinkage` should be enough.

File: clang/lib/Sema/SemaDecl.cpp [hunk 7]
```diff
@@ -12370,6 +12379,17 @@ void Sema::CheckMain(FunctionDecl* FD, const DeclSpec& DS) {
...
+bool Sema::CheckLinkageSpecification(DeclContext *DC, Decl *D) {
+  // [basic.start.main] p2
+  //   The main function shall not be declared with a linkage-specification.
+  if (DC->isExternCContext()) {
+    Diag(D->getLocation(), diag::err_main_invalid_linkage_specification);
+    D->setInvalidDecl();
+    return true;
+  }
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (Sirraide) [reply clang/lib/Sema/SemaDecl.cpp:12391]: Currently, it’s only used in one place (the use where we check variable declarations should not be there I think), so I’d just inline it.

File: clang/test/SemaCXX/linkage3.cpp [hunk 8]
```diff
@@ -0,0 +1,5 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -pedantic-errors %s
+
```
Comment (Sirraide) [inline clang/test/SemaCXX/linkage3.cpp:2]: Also add a test for this (at the top level w/o any linkage-spec-decl) if we don’t already have one:
```
int main;
```
which is also invalid; probably in a separate file so it doesn’t complain about redeclarations

File: clang/lib/AST/Decl.cpp [hunk 9]
```diff
@@ -3292,11 +3292,9 @@ bool FunctionDecl::isImmediateFunction() const {
 }
 
 bool FunctionDecl::isMain() const {
-  const TranslationUnitDecl *tunit =
-    dyn_cast<TranslationUnitDecl>(getDeclContext()->getRedeclContext());
-  return tunit &&
-         !tunit->getASTContext().getLangOpts().Freestanding &&
-         isNamed(this, "main");
+  const DeclContext *DC = getDeclContext();
+  return isNamed(this, "main") && !getLangOpts().Freestanding &&
```
[the remark this answers is not in the document]
Comment (Sirraide) [reply clang/lib/AST/Decl.cpp:3296]: I don’t see how they couldn’t be the same seeing as `Decl`’s `getLangOpts()` just gets them from the `ASTContext` either way. I think this change is fine.

File: clang/docs/ReleaseNotes.rst [hunk 10]
```diff
@@ -145,6 +145,8 @@ Improvements to Clang's diagnostics
 
 - -Wdangling-assignment-gsl is enabled by default.
 
+- Clang now diagnoses the use of `main` in `extern` context as invalid according to [basic.start.main] p2. Fixes #GH101512.
```
Comment (Sirraide) [inline clang/docs/ReleaseNotes.rst:148]:
```suggestion
- Clang now diagnoses the use of `main` in an `extern "C"` context as invalid according to [basic.start.main] p2. Fixes #GH101512.
```

**Review by Sirraide — CHANGES_REQUESTED** [review]
This needs some cleanup I think, but it’s going in the right direction.

File: clang/lib/Sema/SemaDecl.cpp [hunk 11]
```diff
@@ -8052,15 +8061,13 @@ NamedDecl *Sema::ActOnVariableDeclarator(
...
-      !getLangOpts().Freestanding && !NewVD->getDescribedVarTemplate()) {
-
+  if (isMainVar(Name, NewVD) && !getLangOpts().Freestanding) {
     // C++ [basic.start.main]p3
     // A program that declares a variable main at global scope is ill-formed.
-    if (getLangOpts().CPlusPlus)
-      Diag(D.getBeginLoc(), diag::err_main_global_variable);
-
+    if (getLangOpts().CPlusPlus) {
+      if (!CheckLinkageSpecification(DC, NewVD))
```
[the remark this answers is not in the document]
Comment (Sirraide) [reply clang/lib/Sema/SemaDecl.cpp:8068]: [[basic.start.main]p1](https://eel.is/c++draft/basic.start.main#3.1):
> a variable `main` that belongs to the global scope, or

File: clang/lib/Sema/SemaDecl.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: @Sirraide Thanks for the feedback. I've moved directly to `CheckMain`

File: clang/lib/Sema/SemaDecl.cpp (hunk 3 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: @Sirraide Changed to `isExternC()`

File: clang/lib/Sema/SemaDecl.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: There is a test to cover this case in [p3](https://github.com/llvm/llvm-project/blob/cc25748d07b0adb1a0223e836e9cae22af150496/clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp#L13)

File: clang/lib/Sema/SemaDecl.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: @Sirraide @MitalAshok could you review the latest changes? Thanks

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 12]
```diff
@@ -990,6 +990,12 @@ def warn_main_redefined : Warning<"variable named 'main' with external linkage "
     "has undefined behavior">, InGroup<Main>;
 def ext_main_used : Extension<
     "referring to 'main' within an expression is a Clang extension">, InGroup<Main>;
+def ext_main_invalid_linkage_specification : ExtWarn<
+  "'main' should not be "
+  "'extern \"%select{C|C++}0\"'">, InGroup<Main>;
+def err_main_invalid_linkage_specification : Error<
+  "'main' should not be "
+  "'extern \"%select{C|C++}0\"'">;
```
Comment (Sirraide) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:998]: Thinking about this again, I think making these an error by default is a bit too strong; no-one else does that by default; the most you’ll get is a warning from EDG and a pedantic warning from GCC. I think making this a single `ExtWarn` is fine.

**Review by Sirraide — COMMENTED** [review]
A few minor things, but other than that this is pretty much done I’d say.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 12 above)
Comment (a-tarasyuk) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: @Sirraide Thanks for mentioning about the `.Summary`. I’ll keep it in mind for future fixes. I've added the requested changes.

File: clang/lib/Sema/SemaDecl.cpp [hunk 13]
```diff
@@ -7353,6 +7353,15 @@ void emitReadOnlyPlacementAttrWarning(Sema &S, const VarDecl *VD) {
   }
 }
 
+// Checks if VD is declared at global scope or with C language linkage
```
Comment (Sirraide) [inline clang/lib/Sema/SemaDecl.cpp:7356]:
```suggestion
// Checks if VD is declared at global scope or with C language linkage.
```
Extremely minor nit.

File: clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp [hunk 14]
```diff
@@ -8,6 +8,9 @@
 // RUN: %clang_cc1 -fsyntax-only -verify %s -DTEST8
 // RUN: %clang_cc1 -fsyntax-only -verify %s -DTEST9
 // RUN: %clang_cc1 -fsyntax-only -verify %s -DTEST10 -ffreestanding
+// RUN: %clang_cc1 -fsyntax-only -verify -pedantic-errors %s -DTEST11
+// RUN: %clang_cc1 -fsyntax-only -verify -pedantic-errors %s -DTEST12
+// RUN: %clang_cc1 -fsyntax-only -verify -pedantic-errors %s -DTEST13
```
Comment (Sirraide) [inline clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp:13]: I’d prefer just `-pedantic` or `-Wmain` here so we can test that the warnings are actually warnings as they should be

**Review by Sirraide — APPROVED** [review]
Two more minor changes but LGTM otherwise.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 12 above)
Comment (Sirraide) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: > Is it advisable to use force push, or is it better to keep the commit history in PRs?

I don’t think we have a policy on that officially, and it doesn’t really matter because we do squash+merge anyway, so the commit history for a pr can be as messy as it needs to be. *Personally*, however, I’d prefer it if people *didn’t* use force pushes simply because that way I know what was added and what I haven’t reviewed yet instead of having to go over the entire pr again.

File: clang/lib/Sema/SemaDecl.cpp (hunk 13 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: Fixed

File: clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp (hunk 14 above)
Comment (a-tarasyuk) [reply clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp]: I've changed to `-pedantic`

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 12 above)
Comment (a-tarasyuk) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: @Sirraide thanks for the detailed clarification.

File: clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp [hunk 15]
```diff
@@ -61,6 +64,64 @@ int q(void)
 // expected-no-diagnostics
 int main;
 
+#elif TEST11
+extern "C" {
+  namespace Y {
+    int main; // expected-error {{main cannot be declared as global variable}}
```
Comment (MitalAshok) [inline clang/test/CXX/basic/basic.start/basic.start.main/p3.cpp:70]: These diagnostics could be similar to: https://github.com/llvm/llvm-project/blob/150bf637baad2ba4df6369600b65d897ed9b31a7/clang/include/clang/Basic/DiagnosticSemaKinds.td#L6083-L6087

So `main cannot be declared as a variable %select{in the global scope|with C language linkage}0`.

File: clang/lib/Sema/SemaDecl.cpp [hunk 16]
```diff
@@ -8052,10 +8061,7 @@ NamedDecl *Sema::ActOnVariableDeclarator(
   }
 
   // Special handling of variable named 'main'.
-  if (Name.getAsIdentifierInfo() && Name.getAsIdentifierInfo()->isStr("main") &&
-      NewVD->getDeclContext()->getRedeclContext()->isTranslationUnit() &&
-      !getLangOpts().Freestanding && !NewVD->getDescribedVarTemplate()) {
-
+  if (!getLangOpts().Freestanding && isMainVar(Name, NewVD)) {
     // C++ [basic.start.main]p3
     // A program that declares a variable main at global scope is ill-formed.
```
Comment (MitalAshok) [inline clang/lib/Sema/SemaDecl.cpp:8066]:
```suggestion
    // C++ [basic.start.main]p3:
    //   A program that declares
    //    - a variable main at global scope, or
    //    - an entity named main with C language linkage (in any namespace)
    //   is ill-formed.
```

To facilitate the diagnostic showing whether its a global variable or a C linkage variable, you could make `isMainVar` return an enum like `IMV_NotMain, IMV_GlobalMain, IMV_CLinkageMain`. Or you could inline the `isMainVar` function into here again since there would be three options now. This quote from the standard should be good enough documentation

File: clang/lib/Sema/SemaDecl.cpp [hunk 17]
```diff
@@ -12238,7 +12255,6 @@ void Sema::CheckMain(FunctionDecl* FD, const DeclSpec& DS) {
         << FixItHint::CreateRemoval(DS.getConstexprSpecLoc());
     FD->setConstexprKind(ConstexprSpecKind::Unspecified);
   }
```
Comment (MitalAshok) [inline clang/lib/Sema/SemaDecl.cpp:12257]:
```suggestion
  }

```

(Unnecessary diff)

**Review by MitalAshok — APPROVED** [review]
Diagnostic changes are a minor nit, LGTM otherwise

File: clang/lib/Sema/SemaDecl.cpp (hunk 16 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: @MitalAshok Thanks for the feedback. I've added `enum`.

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: @MitalAshok Thanks for the reviews. I’ve added requested changes. 

@MitalAshok @Sirraide If there’s no further feedback, could you please proceed with the merge since I don’t have access? Thanks.

File: clang/lib/Sema/SemaDecl.cpp [hunk 18]
```diff
@@ -7353,13 +7353,26 @@ void emitReadOnlyPlacementAttrWarning(Sema &S, const VarDecl *VD) {
...
-  return Name.getAsIdentifierInfo() &&
-         Name.getAsIdentifierInfo()->isStr("main") &&
-         !VD->getDescribedVarTemplate() &&
-         (VD->getDeclContext()->getRedeclContext()->isTranslationUnit() ||
-          VD->isExternC());
+enum MainVarClassificationKind {
+  IMV_NotMain,
+  IMV_GlobalMain,
+  IMV_CLinkageMain,
+};
```
Comment (Sirraide) [inline clang/lib/Sema/SemaDecl.cpp:7360]: I don’t think this needs to exist:
1. We’re not even using all of these enumerators.
2. We test for them in exactly one place.
3. The test in question ends up being more complex as a result.

It was fine the way it was before, so I’d suggest you just revert this commit.

(Also, for future reference, we prefer scoped enumerations these days.)

**Review by Sirraide — CHANGES_REQUESTED** [review]

File: clang/lib/Sema/SemaDecl.cpp (hunk 18 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp]: @Sirraide Oke, reverted.

**Review by Sirraide — APPROVED** [review]
LGTM.

I’ll wait a bit to see if anyone else wants to comment on this since we’ve gone back and forth over a few things w/ this pr, but if not, I’ll merge this sometime later.

**Comment by a-tarasyuk** [comment]
@Sirraide thanks

**Comment by alexfh** [comment]
Hi @a-tarasyuk, it looks like this patch introduces a couple of problems. I guess, this boils down to marking the declaration of main() invalid in case it's declared with `extern "C"`, but I may be wrong here.

The first problem is a clang crash: https://gcc.godbolt.org/z/dWETErhzM

The second one is a change of behavior of code that declares `main()` `extern "C"` (see my comment https://github.com/llvm/llvm-project/commit/37ec6e5f12afe4a37872bf28f280423696f39019#r145380228). This one still needs some work to create a standalone repro for, but I suspect that both issues are related to marking the declaration invalid.

The problem with both cases is that in some codebases the non-trivial amount of cleanup is necessary for this, but Clang doesn't allow to retain the old behavior by just ignoring the diagnostic (`-Wno-main`). Even if the diagnostic is not issued, the issues above still manifest.

Please fix this ASAP or revert the patch, if it requires non-trivial / controversial changes or further discussion.

Thanks!

File: clang/lib/Sema/SemaDecl.cpp [hunk 19]
```diff
@@ -12210,7 +12220,18 @@ bool Sema::CheckFunctionDeclaration(Scope *S, FunctionDecl *NewFD,
...
+void Sema::CheckMain(FunctionDecl *FD, const DeclSpec &DS) {
+  // [basic.start.main]p3
+  //    The main function shall not be declared with a linkage-specification.
+  if (FD->isExternCContext() ||
+      (FD->isExternCXXContext() &&
+       FD->getDeclContext()->getRedeclContext()->isTranslationUnit())) {
+    Diag(FD->getLocation(), diag::ext_main_invalid_linkage_specification)
+        << FD->getLanguageLinkage();
+    FD->setInvalidDecl();
+    return;
```
Comment (MitalAshok) [inline clang/lib/Sema/SemaDecl.cpp:12232]:
```suggestion
```

@alexfh I think these lines are probably what's causing the crash (if it's allowed as an extension, it should not be marked invalid, and we need to diagnose/check the other stuff if we have `-Wno-main`).

Either way, could you open a new GitHub issue for visibility?

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp:12232]: @alexfh I’ll look into it today. Thanks for bringing this additional case to my attention.

Comment (alexfh) [reply clang/lib/Sema/SemaDecl.cpp:12232]: Filed https://github.com/llvm/llvm-project/issues/104570

Comment (alexfh) [reply clang/lib/Sema/SemaDecl.cpp:12232]: The fix is in review (#104594).
